File: ScriptsToProcess/NEW BATCH TO PROCESS/GridPattern.py
import Rhino
import math
import System
from Rhino.Geometry import Rectangle3d, Plane, Point3d, Curve, Transform
from Rhino.Geometry.Intersect import Intersection

# Update component info
ghenv.Component.Name = "Grid Pattern Generator and Trimmer"
ghenv.Component.NickName = "GridPattern"
ghenv.Component.Message = ""

# Initialize debug messages list globally
debug_messages = []

def debug_print(msg):
    # Only add to out parameter, not to component message
    print(msg)
    debug_messages.append(msg)

class GridCell:
    def __init__(self):
        self.curve = None
        self.center = None
        self.corners = []
    
    def get_test_points(self):
        return self.corners + [self.center] if self.center else self.corners
    
    def transform(self, xform):
        if self.curve:
            self.curve.Transform(xform)
        if self.center:
            self.center.Transform(xform)
        for i in range(len(self.corners)):
            self.corners[i].Transform(xform)

class GridGenerator:
    def __init__(self, boundary, cell_width, cell_height, grid_type="rectangular", plane=Plane.WorldXY):
        self.boundary = boundary
        self.cell_width = cell_width
        self.cell_height = cell_height
        self.grid_type = grid_type
        self.plane = plane
        
        # Create transforms between world and local coordinates
        self.to_local = Transform.PlaneToPlane(self.plane, Plane.WorldXY)
        self.to_world = Transform.PlaneToPlane(Plane.WorldXY, self.plane)
        
        # Transform boundary to local coordinates for calculations
        self.local_boundary = boundary.DuplicateCurve()
        self.local_boundary.Transform(self.to_local)
        
        # Get bounding box in local coordinates
        self.bbox = self.local_boundary.GetBoundingBox(True)
        debug_print(f"Local boundary box: Min({self.bbox.Min}), Max({self.bbox.Max})")
    
    def create_rectangular_cell(self, x, y, offset=0):
        cell = GridCell()
        # Apply offset to x-coordinate
        adjusted_x = x + offset
        
        cell.curve = Rectangle3d(
            Plane.WorldXY, 
            Point3d(adjusted_x, y, 0), 
            Point3d(adjusted_x + self.cell_width, y + self.cell_height, 0)
        ).ToNurbsCurve()
        cell.center = Point3d(adjusted_x + self.cell_width/2, y + self.cell_height/2, 0)
        cell.corners = [
            Point3d(adjusted_x, y, 0),
            Point3d(adjusted_x + self.cell_width, y, 0),
            Point3d(adjusted_x + self.cell_width, y + self.cell_height, 0),
            Point3d(adjusted_x, y + self.cell_height, 0)
        ]
        return cell
    
    def create_hexagonal_cell(self, center_x, center_y):
        cell = GridCell()
        radius = self.cell_width / 2
        corners = []
        for i in range(6):
            angle = i * math.pi / 3
            x = center_x + radius * math.cos(angle)
            y = center_y + radius * math.sin(angle)
            corners.append(Point3d(x, y, 0))
        
        cell.corners = corners
        cell.center = Point3d(center_x, center_y, 0)
        cell.curve = Curve.CreateInterpolatedCurve(corners + [corners[0]], 1)
        return cell
    
    def create_triangular_cell(self, base_x, base_y, inverted=False):
        cell = GridCell()
        height = self.cell_width * math.sqrt(3)/2
        
        if not inverted:
            corners = [
                Point3d(base_x, base_y, 0),
                Point3d(base_x + self.cell_width, base_y, 0),
                Point3d(base_x + self.cell_width/2, base_y + height, 0)
            ]
        else:
            corners = [
                Point3d(base_x + self.cell_width/2, base_y, 0),
                Point3d(base_x + self.cell_width, base_y + height, 0),
                Point3d(base_x, base_y + height, 0)
            ]
        
        cell.corners = corners
        cell.center = Point3d(sum(p.X for p in corners)/3, sum(p.Y for p in corners)/3, 0)
        cell.curve = Curve.CreateInterpolatedCurve(corners + [corners[0]], 1)
        return cell
    
    def generate_cells(self):
        debug_print(f"Generating {self.grid_type} grid with cell size: {self.cell_width} x {self.cell_height}")
        
        cells = []
        if self.grid_type in ["rectangular", "offset_rectangular"]:
            x = math.floor(self.bbox.Min.X / self.cell_width) * self.cell_width
            while x < self.bbox.Max.X:
                y = math.floor(self.bbox.Min.Y / self.cell_height) * self.cell_height
                row_index = int((y - self.bbox.Min.Y) / self.cell_height)
                
                while y < self.bbox.Max.Y:
                    # For offset_rectangular, apply 0.5 * cell_width offset on alternate rows
                    offset = (0.5 * self.cell_width) if (self.grid_type == "offset_rectangular" and row_index % 2 == 1) else 0
                    cell = self.create_rectangular_cell(x, y, offset)
                    cells.append(cell)
                    y += self.cell_height
                    row_index += 1
                x += self.cell_width
                
        elif self.grid_type == "hexagonal":
            h_spacing = self.cell_width * 3/4
            v_spacing = self.cell_width * math.sqrt(3)/2
            x = math.floor(self.bbox.Min.X / h_spacing) * h_spacing
            row_count = 0
            while x < self.bbox.Max.X:
                y = math.floor(self.bbox.Min.Y / v_spacing) * v_spacing
                offset = (v_spacing/2) if row_count % 2 else 0
                while y < self.bbox.Max.Y:
                    cell = self.create_hexagonal_cell(x, y + offset)
                    cells.append(cell)
                    y += v_spacing
                row_count += 1
                x += h_spacing
                
        elif self.grid_type == "triangular":
            height = self.cell_width * math.sqrt(3)/2
            x = math.floor(self.bbox.Min.X / self.cell_width) * self.cell_width
            while x < self.bbox.Max.X:
                y = math.floor(self.bbox.Min.Y / height) * height
                while y < self.bbox.Max.Y:
                    cells.append(self.create_triangular_cell(x, y, False))
                    cells.append(self.create_triangular_cell(x, y, True))
                    y += height
                x += self.cell_width
                
        debug_print(f"Generated {len(cells)} cells")
        return cells

def point_containment_test(point, curve, plane, tolerance=0.001):
    containment = curve.Contains(point, plane, tolerance)
    return containment == Rhino.Geometry.PointContainment.Inside

try:
    # Clear previous debug messages
    debug_messages = []
    debug_print("Starting script execution")
    
    # Input validation
    if not boundary:
        raise ValueError("No boundary curve provided")
    
    if isinstance(boundary, list):
        boundaries = [b.ToNurbsCurve() if isinstance(b, Rhino.Geometry.PolylineCurve) else b for b in boundary]
    else:
        boundaries = [boundary.ToNurbsCurve() if isinstance(boundary, Rhino.Geometry.PolylineCurve) else boundary]
    
    for b in boundaries:
        if not b.IsClosed:
            raise ValueError("All boundaries must be closed curves")
    
    grid_type = grid_type if grid_type else "rectangular"
    x_dim = x_dim if x_dim else 1.0
    y_dim = y_dim if y_dim else 1.0
    origin_plane = origin_plane if origin_plane else Plane.WorldXY
    
    debug_print(f"Processing with parameters: grid_type={grid_type}, x_dim={x_dim}, y_dim={y_dim}")
    
    # Initialize outputs with better names
    all_grid_cells = []  # instead of a
    cell_trim_status = []  # instead of b
    complete_cell_count = 0  # instead of c
    trimmed_cell_count = 0  # instead of d
    complete_cells_only = []  # instead of e
    
    for boundary_curve in boundaries:
        debug_print(f"Processing boundary: {boundary_curve}")
        
        # Generate grid
        grid = GridGenerator(boundary_curve, x_dim, y_dim, grid_type, origin_plane)
        cells = grid.generate_cells()
        
        # Process cells
        for cell in cells:
            # Transform cell to world coordinates for intersection testing
            cell.transform(grid.to_world)
            test_points = cell.get_test_points()
            points_inside = [point_containment_test(pt, boundary_curve, origin_plane) for pt in test_points]
            
            intersection_events = Intersection.CurveCurve(cell.curve, boundary_curve, 0.001, 0.001)
            has_intersection = intersection_events is not None and intersection_events.Count > 0
            
            if all(points_inside):
                all_grid_cells.append(cell.curve)
                complete_cells_only.append(cell.curve)
                cell_trim_status.append(0)
                complete_cell_count += 1
            elif any(points_inside) or has_intersection:
                intersect_curves = Rhino.Geometry.Curve.CreateBooleanIntersection(cell.curve, boundary_curve)
                if intersect_curves and len(intersect_curves) > 0:
                    for trimmed_curve in intersect_curves:
                        all_grid_cells.append(trimmed_curve)
                        cell_trim_status.append(1)
                        trimmed_cell_count += 1
    
    debug_print(f"Final counts - Complete cells: {complete_cell_count}, Trimmed cells: {trimmed_cell_count}")
    
    # Update component message to be cleaner
    ghenv.Component.Message = f"{grid_type.capitalize()} Grid\n{complete_cell_count} complete | {trimmed_cell_count} trimmed"
    
    # Set outputs with your preferred names
    allTiles = all_grid_cells
    panelStatus = cell_trim_status
    fullCount = complete_cell_count
    trimCount = trimmed_cell_count
    fullTiles = complete_cells_only
    
    # Assign to component outputs
    a = allTiles
    b = panelStatus
    c = fullCount
    d = trimCount
    e = fullTiles
    
    # Combine all debug messages into the out parameter
    out = "\n".join(debug_messages)

except Exception as e:
    error_msg = f"Error: {str(e)}\nType: {type(e)}"
    print(error_msg)
    ghenv.Component.Message = "Error occurred"
    a, b, c, d, e = None, None, None, None, None
    out = error_msg

File: Components/GridPattern.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Grasshopper.Kernel;
using Grasshopper.Kernel.Data;
using Grasshopper.Kernel.Types;
using Rhino.Geometry;
using Rhino.Geometry.Intersect;

namespace Enzyme.Components
{
    public class GridPattern : GH_Component
    {
        public GridPattern()
          : base("Grid Pattern Generator and Trimmer", "GridPattern",
              "",
              "Enzyme", "Pattern")
        {
        }

        public override Guid ComponentGuid => new Guid("D7A50BEF-4309-4C41-BCA6-455AB0B2C471");

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("GridPattern.png");
            }
        }

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 3, 0.0, 2.0, 1.0, 330, -20);
                Enzyme.Utils.AutoWireHelper.WireSlider(this, document, 4, 0.0, 2.0, 1.0, 330, 20);
                Enzyme.Utils.AutoWireHelper.WireOutputParam(this, document, 0, "curve", 220, -23);
                Enzyme.Utils.AutoWireHelper.WireOutputParam(this, document, 4, "curve", 220, 22);
            }
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddCurveParameter("Boundary", "Boundary", "Closed boundary curve(s)", GH_ParamAccess.tree);
            pManager.AddPlaneParameter("Origin Plane", "origin_plane", "Grid base plane", GH_ParamAccess.item, Plane.WorldXY);
            pManager.AddTextParameter("Grid Type", "grid_type", "Grid type: rectangular, offset_rectangular, hexagonal, triangular", GH_ParamAccess.item, "rectangular");
            pManager.AddNumberParameter("Cell Width", "x_dim", "Cell width", GH_ParamAccess.item, 1.0);
            pManager.AddNumberParameter("Cell Height", "y_dim", "Cell height", GH_ParamAccess.item, 1.0);
            
            pManager[1].Optional = true;
            pManager[2].Optional = true;
            pManager[3].Optional = true;
            pManager[4].Optional = true;
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddCurveParameter("All Tiles", "a", "All grid tiles", GH_ParamAccess.list);
            pManager.AddIntegerParameter("Tile Status", "b", "0 = full, 1 = trimmed", GH_ParamAccess.list);
            pManager.AddIntegerParameter("Full Count", "c", "Number of complete cells", GH_ParamAccess.item);
            pManager.AddIntegerParameter("Trim Count", "d", "Number of trimmed cells", GH_ParamAccess.item);
            pManager.AddCurveParameter("Full Tiles", "e", "Tiles fully inside the boundary", GH_ParamAccess.list);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            if (!DA.GetDataTree(0, out GH_Structure<GH_Curve> boundaryTree))
                return;

            Plane originPlane = Plane.WorldXY;
            DA.GetData(1, ref originPlane);

            string gridType = "rectangular";
            DA.GetData(2, ref gridType);

            double x_dim = 1.0;
            DA.GetData(3, ref x_dim);

            double y_dim = 1.0;
            DA.GetData(4, ref y_dim);

            List<Curve> allTiles = new List<Curve>();
            List<int> panelStatus = new List<int>();
            int fullCount = 0;
            int trimCount = 0;
            List<Curve> fullTiles = new List<Curve>();

            List<Curve> boundaries = new List<Curve>();
            foreach (GH_Path path in boundaryTree.Paths)
            {
                var branch = boundaryTree.get_Branch(path);
                foreach (GH_Curve ghCurve in branch)
                {
                    if (ghCurve != null && ghCurve.Value != null)
                    {
                        Curve crv = ghCurve.Value;
                        if (crv is PolylineCurve)
                            crv = crv.ToNurbsCurve();
                        
                        if (crv.IsClosed)
                            boundaries.Add(crv);
                        else
                            AddRuntimeMessage(GH_RuntimeMessageLevel.Warning, "Boundary must be a closed curve");
                    }
                }
            }

            if (boundaries.Count == 0)
            {
                AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "No valid closed boundaries provided.");
                return;
            }

            foreach (Curve boundary in boundaries)
            {
                // Create transforms
                Transform toLocal = Transform.PlaneToPlane(originPlane, Plane.WorldXY);
                Transform toWorld = Transform.PlaneToPlane(Plane.WorldXY, originPlane);

                Curve localBoundary = boundary.DuplicateCurve();
                localBoundary.Transform(toLocal);
                BoundingBox bbox = localBoundary.GetBoundingBox(true);

                List<GridCell> cells = GenerateCells(bbox, x_dim, y_dim, gridType);

                foreach (GridCell cell in cells)
                {
                    cell.Transform(toWorld);
                    List<Point3d> testPoints = cell.GetTestPoints();

                    bool[] pointsInside = testPoints.Select(pt => PointContainmentTest(pt, boundary, originPlane)).ToArray();
                    bool allInside = pointsInside.All(p => p);
                    bool anyInside = pointsInside.Any(p => p);

                    CurveIntersections intersections = Intersection.CurveCurve(cell.Curve, boundary, 0.001, 0.001);
                    bool hasIntersection = intersections != null && intersections.Count > 0;

                    if (allInside)
                    {
                        allTiles.Add(cell.Curve);
                        fullTiles.Add(cell.Curve);
                        panelStatus.Add(0);
                        fullCount++;
                    }
                    else if (anyInside || hasIntersection)
                    {
                        Curve[] intersectCurves = Curve.CreateBooleanIntersection(cell.Curve, boundary, 0.001);
                        if (intersectCurves != null && intersectCurves.Length > 0)
                        {
                            foreach (Curve trimmedCurve in intersectCurves)
                            {
                                allTiles.Add(trimmedCurve);
                                panelStatus.Add(1);
                                trimCount++;
                            }
                        }
                    }
                }
            }

            string capGridType = gridType.Length > 0 ? char.ToUpper(gridType[0]) + gridType.Substring(1).ToLower() : gridType;
            Message = $"{capGridType} Grid\n{fullCount} complete | {trimCount} trimmed";

            DA.SetDataList(0, allTiles);
            DA.SetDataList(1, panelStatus);
            DA.SetData(2, fullCount);
            DA.SetData(3, trimCount);
            DA.SetDataList(4, fullTiles);
        }

        private class GridCell
        {
            public Curve Curve { get; set; }
            public Point3d Center { get; set; }
            public List<Point3d> Corners { get; set; } = new List<Point3d>();

            public List<Point3d> GetTestPoints()
            {
                List<Point3d> pts = new List<Point3d>(Corners);
                if (Center.IsValid) pts.Add(Center);
                return pts;
            }

            public void Transform(Transform xform)
            {
                if (Curve != null) Curve.Transform(xform);
                Center = new Point3d(Center);
                Point3d c = Center;
                c.Transform(xform);
                Center = c;
                for (int i = 0; i < Corners.Count; i++)
                {
                    Point3d p = Corners[i];
                    p.Transform(xform);
                    Corners[i] = p;
                }
            }
        }

        private List<GridCell> GenerateCells(BoundingBox bbox, double cellWidth, double cellHeight, string gridType)
        {
            List<GridCell> cells = new List<GridCell>();

            if (gridType == "rectangular" || gridType == "offset_rectangular")
            {
                double x = Math.Floor(bbox.Min.X / cellWidth) * cellWidth;
                while (x < bbox.Max.X)
                {
                    double y = Math.Floor(bbox.Min.Y / cellHeight) * cellHeight;
                    int rowIndex = (int)Math.Round((y - bbox.Min.Y) / cellHeight);

                    while (y < bbox.Max.Y)
                    {
                        double offset = (gridType == "offset_rectangular" && rowIndex % 2 == 1) ? 0.5 * cellWidth : 0;
                        cells.Add(CreateRectangularCell(x, y, offset, cellWidth, cellHeight));
                        y += cellHeight;
                        rowIndex++;
                    }
                    x += cellWidth;
                }
            }
            else if (gridType == "hexagonal")
            {
                double hSpacing = cellWidth * 3.0 / 4.0;
                double vSpacing = cellWidth * Math.Sqrt(3) / 2.0;
                double x = Math.Floor(bbox.Min.X / hSpacing) * hSpacing;
                int rowCount = 0;
                while (x < bbox.Max.X)
                {
                    double y = Math.Floor(bbox.Min.Y / vSpacing) * vSpacing;
                    double offset = (rowCount % 2 != 0) ? vSpacing / 2.0 : 0;
                    while (y < bbox.Max.Y)
                    {
                        cells.Add(CreateHexagonalCell(x, y + offset, cellWidth));
                        y += vSpacing;
                    }
                    rowCount++;
                    x += hSpacing;
                }
            }
            else if (gridType == "triangular")
            {
                double height = cellWidth * Math.Sqrt(3) / 2.0;
                double x = Math.Floor(bbox.Min.X / cellWidth) * cellWidth;
                while (x < bbox.Max.X)
                {
                    double y = Math.Floor(bbox.Min.Y / height) * height;
                    while (y < bbox.Max.Y)
                    {
                        cells.Add(CreateTriangularCell(x, y, false, cellWidth, height));
                        cells.Add(CreateTriangularCell(x, y, true, cellWidth, height));
                        y += height;
                    }
                    x += cellWidth;
                }
            }

            return cells;
        }

        private GridCell CreateRectangularCell(double x, double y, double offset, double cellWidth, double cellHeight)
        {
            GridCell cell = new GridCell();
            double adjustedX = x + offset;
            cell.Curve = new Rectangle3d(Plane.WorldXY, new Point3d(adjustedX, y, 0), new Point3d(adjustedX + cellWidth, y + cellHeight, 0)).ToNurbsCurve();
            cell.Center = new Point3d(adjustedX + cellWidth / 2, y + cellHeight / 2, 0);
            cell.Corners = new List<Point3d> {
                new Point3d(adjustedX, y, 0),
                new Point3d(adjustedX + cellWidth, y, 0),
                new Point3d(adjustedX + cellWidth, y + cellHeight, 0),
                new Point3d(adjustedX, y + cellHeight, 0)
            };
            return cell;
        }

        private GridCell CreateHexagonalCell(double centerX, double centerY, double cellWidth)
        {
            GridCell cell = new GridCell();
            double radius = cellWidth / 2.0;
            List<Point3d> corners = new List<Point3d>();
            for (int i = 0; i < 6; i++)
            {
                double angle = i * Math.PI / 3.0;
                corners.Add(new Point3d(centerX + radius * Math.Cos(angle), centerY + radius * Math.Sin(angle), 0));
            }
            cell.Corners = corners;
            cell.Center = new Point3d(centerX, centerY, 0);
            
            List<Point3d> crvPts = new List<Point3d>(corners);
            crvPts.Add(corners[0]);
            cell.Curve = Curve.CreateInterpolatedCurve(crvPts, 1);
            return cell;
        }

        private GridCell CreateTriangularCell(double baseX, double baseY, bool inverted, double cellWidth, double height)
        {
            GridCell cell = new GridCell();
            List<Point3d> corners = new List<Point3d>();
            if (!inverted)
            {
                corners.Add(new Point3d(baseX, baseY, 0));
                corners.Add(new Point3d(baseX + cellWidth, baseY, 0));
                corners.Add(new Point3d(baseX + cellWidth / 2.0, baseY + height, 0));
            }
            else
            {
                corners.Add(new Point3d(baseX + cellWidth / 2.0, baseY, 0));
                corners.Add(new Point3d(baseX + cellWidth, baseY + height, 0));
                corners.Add(new Point3d(baseX, baseY + height, 0));
            }
            cell.Corners = corners;
            cell.Center = new Point3d(corners.Sum(p => p.X) / 3.0, corners.Sum(p => p.Y) / 3.0, 0);

            List<Point3d> crvPts = new List<Point3d>(corners);
            crvPts.Add(corners[0]);
            cell.Curve = Curve.CreateInterpolatedCurve(crvPts, 1);
            return cell;
        }

        private bool PointContainmentTest(Point3d point, Curve curve, Plane plane, double tolerance = 0.001)
        {
            return curve.Contains(point, plane, tolerance) == PointContainment.Inside;
        }
    }
}
